src/H5FDprivate.h

```diff
@@ -39,6 +39,220 @@
+ *              lower file and is therefore about to be removed from the
+ *              metadata file
+ *
+ * garbage: `true` if the entry is marked for garbage collection and is
+ *              thus invalid.
+ *
+ *              For n the number of entries, deleting an entry is O(n).
+ *              H5PB_dest() deletes all entries.  Instead of deleting
+ *              entries one-by-one at O(n^2) cost, H5PB_dest() marks
+ *              each disused entry for garbage collection and sweeps all
+ *              entries up before it is done.
+ *
+ *----------------------------------------------------------------------------
+ */
+typedef struct H5FD_vfd_swmr_idx_entry_t {
+    uint64_t hdf5_page_offset;
+    uint64_t md_file_page_offset;
+    uint32_t length;
+    uint32_t checksum;
+    void *   entry_ptr;
+    uint64_t tick_of_last_change;
+    hbool_t  clean;
+    uint64_t tick_of_last_flush;
+    uint64_t delayed_flush;
+    bool     moved_to_lower_file;
+    bool     garbage;
+} H5FD_vfd_swmr_idx_entry_t;
+
+/*
+ *  tick_num:       Sequence number of the current tick.
+ *                  Initialized to zero on file creation/open, and incremented
+ *                  by the VFD SWMR writer at the end of each tick.
+ *  num_entries:    The number of entries in the index.
+ *  entries:        The array of index entries
+ */
+typedef struct H5FD_vfd_swmr_md_index {
+    uint64_t                   tick_num;
+    uint32_t                   num_entries;
+    H5FD_vfd_swmr_idx_entry_t *entries;
+} H5FD_vfd_swmr_md_index;
+
+/*
+ *  fs_page_size:   Size of pages in both the HDF5 file and the metadata file IN BYTES
+ *  tick_num:       Sequence number of the current tick.
+ *                  Initialized to zero on file creation/open, and incremented by the
+ *                  VFD SWMR writer at the end of each tick.
+ *  index_offset:   The offset of the current metadata file index in the metadata file
+ *                  IN BYTES.
+ *  index_length:   The length of the current metadata file index IN BYTES.
+ */
+typedef struct H5FD_vfd_swmr_md_header {
+    uint32_t fs_page_size;
+    uint64_t tick_num;
+    uint64_t index_offset;
+    size_t   index_length;
+} H5FD_vfd_swmr_md_header;
+
+/* Lookup the shadow-index entry corresponding to page number `target_page`
+ * in the HDF5 file and return it.  If there is no match, return NULL.
+ *
+ * The lookup is performed by binary search on the `nentries` shadow index
+ * entries at `idx`.  The entries must be sorted by their offset in the
+ * HDF5 file.  Each entry must have a unique HDF5 file offset.
+ *
+ * If `reuse_garbage` is true, then entries marked for garbage collection
+ * are eligible search results.  Return NULL if a matching entry is
+ * found, but the entry is marked for garbage collection and `reuse_garbage`
+ * is false.
+ */
+static inline H5FD_vfd_swmr_idx_entry_t *
+H5FD_vfd_swmr_pageno_to_mdf_idx_entry(H5FD_vfd_swmr_idx_entry_t *idx, uint32_t nentries, uint64_t target_page,
```

Comment:
Please use StdC bsearch() routine instead of re-coding binary search algorithm from scratch.